File: House Robber 2/Memoization.py
def HouseRobber(size,arr,dp):
    if size == 0: return arr[0]
    if size < 0: return 0
    if dp[size] != -1: return dp[size]

    pick = arr[size] + HouseRobber(size-2,arr,dp)
    nonPick = HouseRobber(size-1,arr,dp)

    dp[size] = max(pick,nonPick)
    return dp[size]

if __name__ == '__main__':
    arr = [1,2,3,1]
    size =  len(arr)

    arr1 = arr[:size-1]
    arr2 = arr[1:]

    dp1 = [-1 for _ in range(size-1)]
    dp2 = [-1 for _ in range(size-1)]

    print( max(HouseRobber(size-2,arr1,dp1) , HouseRobber(size-2,arr2,dp2)) )

File: House Robber 2/Memoization.java

import java.util.Arrays;


class Memoization {
     public static int maxRecursion(int size,int arr[],int[] dp) {
        if(size < 0) return 0;
        if(dp[size] != -1) return dp[size];
        if(size == 0) return arr[size];

        int pick = arr[size] + maxRecursion(size-2,arr,dp);
        int nonPick = maxRecursion(size-1,arr,dp);

        return dp[size] =  Math.max(pick,nonPick);
    }

    public static void main(String[] args) {
        int[] nums = {2,2,3};
        int size = nums.length;   

        int arr1[] = new int[size-1];
        int arr2[] = new int[size-1];
        int dp1[] = new int[size-1];
        int dp2[] = new int[size-1];

        Arrays.fill(dp1,-1);
        Arrays.fill(dp2,-1);

        int k = 0;
        for(int i = 0 ; i < size; i ++ ) {
            if( i > 0) {
                arr2[k] = nums[i];
                k += 1;
            }

            if( i < size-1) arr1[i] = nums[i];
        }

        System.out.println(Math.max( maxRecursion(size-2,arr1,dp1), maxRecursion(size-2,arr2,dp2)) );

    }
}